### PR TITLE
Enable Knative Kafka w/o upstream e2e tests and w/o smoke tests

### PR DESCRIPTION
**Pull request by aliok**
- Enable controller and webhook
- No smoke tests yet (they are in progress: https://github.com/lberk/serverless-operator/tree/knativekafkae2e)
- No upstream e2e tests enabled yet (we have [SRVKE-590](https://issues.redhat.com/browse/SRVKE-590) to enable the e2e test)
- Serverless operator e2e tests (deployment shape check) is added
- Use our images as image override is to be added later (for now we need our own images, due to patches - with the CSV override we go back to `gcr.io` for the upstream manifest)
- Had to play with `go mod` because of https://github.com/openshift-knative/serverless-operator/pull/581#pullrequestreview-510827292

### CODE REVIEW

**Comment by aliok** [comment]
Compile error:
```
test/v1alpha1/knativekafka.go:17:3: cannot use "github.com/openshift-knative/serverless-operator/test/vendor/k8s.io/apimachinery/pkg/apis/meta/v1".ObjectMeta literal (type "github.com/openshift-knative/serverless-operator/test/vendor/k8s.io/apimachinery/pkg/apis/meta/v1".ObjectMeta) as type "github.com/openshift-knative/serverless-operator/knative-operator/vendor/k8s.io/apimachinery/pkg/apis/meta/v1".ObjectMeta in field value
```

File: knative-operator/deploy/resources/knativekafka/kafkachannel-v0.17.1.yaml
Comment (matzew) [inline knative-operator/deploy/resources/knativekafka/kafkachannel-v0.17.1.yaml:455]: @markusthoemmes  until we have the CSV override we need those, since we run here from a patch.

File: openshift/ci-operator/source-image/Dockerfile
Comment (markusthoemmes) [inline openshift/ci-operator/source-image/Dockerfile:8]: Why is this needed and wasn't before?

File: README.md
Comment (markusthoemmes) [inline README.md:51]: Is this accurate? Seems like kafka would not be installed by this.

File: knative-operator/deploy/resources/knativekafka/kafkachannel-v0.17.1.yaml
Comment (markusthoemmes) [reply knative-operator/deploy/resources/knativekafka/kafkachannel-v0.17.1.yaml:455]: Wondering why we would ship the CSV overrides after this? I would've expected a PR before this to ship them.

File: olm-catalog/serverless-operator/manifests/operator_v1alpha1_knativekafka_crd.yaml
Comment (markusthoemmes) [inline olm-catalog/serverless-operator/manifests/operator_v1alpha1_knativekafka_crd.yaml:1]: Should this be the v1 based CRD?

File: templates/csv.yaml
Comment (markusthoemmes) [inline templates/csv.yaml:88]: Have we added that? I recall a discussion where we did decide not to add it.

File: test/e2e-tests.sh
Comment (markusthoemmes) [inline test/e2e-tests.sh:26]: How's this supposed to work with the test target we added? :thinking:

File: test/v1alpha1/knativekafka.go
Comment (markusthoemmes) [inline test/v1alpha1/knativekafka.go:49]: Ugh.. Is this because we don't have a client generated for KnativeKafka?

File: knative-operator/deploy/resources/knativekafka/kafkachannel-v0.17.1.yaml
Comment (aliok) [reply knative-operator/deploy/resources/knativekafka/kafkachannel-v0.17.1.yaml:455]: tldr; Image override requires the CSV to have the changes and we didn't want to touch the CSV until we actually enable the controller.

Detail:
--------
CSV override doesn't work as is for Knative Kafka components. For eventing and serving, what we do is we just collect environment vars and pass them to upstream CR. Upstream operator will modify the deployments.
For Kafka components though, we need to modify the deployments ourselves in S-O. 

Another change is, there are some naming clashes. `DISPATCHER_IMAGE` env var is used both in eventing and in knative kafka components. I needed to add "scoping" for that (we once discussed about that).

So, these result in big changes.
- We didn't want to include them in this PR
- We also didn't want to do these in advance because of the "don't do it until you need it" principle (I made up the name but that's the idea I got from all the discussions in #serverless-dev)

File: openshift/ci-operator/source-image/Dockerfile
Comment (aliok) [reply openshift/ci-operator/source-image/Dockerfile:8]: In Knative Kafka, we have a problem that we aren't able to write anything to `/go/src/knative.dev/eventing-contrib`.
This is required for tests.

cc @mgencur

File: templates/csv.yaml
Comment (aliok) [reply templates/csv.yaml:88]: +1 good point. gotta remove that.

File: knative-operator/deploy/resources/knativekafka/kafkachannel-v0.17.1.yaml
Comment (markusthoemmes) [reply knative-operator/deploy/resources/knativekafka/kafkachannel-v0.17.1.yaml:455]: Got it, wfm.

File: test/v1alpha1/knativekafka.go
Comment (aliok) [reply test/v1alpha1/knativekafka.go:49]: yeah. I needed to learn all these ugly crap that's handled by operator-sdk OR knative-pkg

File: olm-catalog/serverless-operator/manifests/operator_v1alpha1_knativekafka_crd.yaml
Comment (aliok) [reply olm-catalog/serverless-operator/manifests/operator_v1alpha1_knativekafka_crd.yaml:1]: +1 on v1 CRD.

Comment (matzew) [reply olm-catalog/serverless-operator/manifests/operator_v1alpha1_knativekafka_crd.yaml:1]: I have moved my "old" PR for CRD_v1 now to Ali's brach:

https://github.com/aliok/serverless-operator/pull/5

See the old reference for the size of the DIFF   - I've now closed my old CRD_v1 PR

Comment (aliok) [reply olm-catalog/serverless-operator/manifests/operator_v1alpha1_knativekafka_crd.yaml:1]: Gonna merge that after CI reports success/fail here

Don't want to interrupt it. I want to see the results

File: test/e2e-tests.sh
Comment (aliok) [reply test/e2e-tests.sh:26]: @markusthoemmes 
we want to progress this as follows:
- see if things are green w/o KnativeKafka serverless e2e tests (aka deployment shape checks)
- once this PR is green, we merge https://github.com/openshift/release/pull/12939 and afterwards we start having a separate `e2e-test-all` target. This runs the deployment shape checks for KnativeKafka too
This way, the existing `make test-e2e` has the same behavior (when done in local) but we use `test-e2e-all` in CI in addition.

@matzew is now on fixing the wrong target name added on master: https://github.com/openshift-knative/serverless-operator/blob/master/Makefile#L40

File: knative-operator/deploy/resources/knativekafka/kafkachannel-v0.17.1.yaml
Comment (markusthoemmes) [reply knative-operator/deploy/resources/knativekafka/kafkachannel-v0.17.1.yaml:455]: It's worth mentioning though that the problem with this is that `master` is not releasable after this PR. Without being able to replace variables via the CSV we can't productize and hence not release.

Comment (matzew) [reply knative-operator/deploy/resources/knativekafka/kafkachannel-v0.17.1.yaml:455]: I agree on this, but there will be a follow up PR of the CSV override soon.

afterwards we are back to a release-able state.

File: README.md
Comment (aliok) [reply README.md]: This is fixed now

File: olm-catalog/serverless-operator/manifests/operator_v1alpha1_knativekafka_crd.yaml
Comment (aliok) [reply olm-catalog/serverless-operator/manifests/operator_v1alpha1_knativekafka_crd.yaml:1]: Let's do this separately. I had trouble doing this. I received strange OLM errors. Perhaps unrelated.
Doing it in a separate PR makes more sense anyway since we already have the CRD with v1beta1 in master and that change is not in this PR's scope

File: templates/csv.yaml
Comment (aliok) [reply templates/csv.yaml:88]: I removed this

File: test/e2e-tests.sh
Comment (aliok) [reply test/e2e-tests.sh:26]: This should be good now. Prow is configured already

**Comment by aliok** [comment]
@markusthoemmes I did rebase and adressed the comments.

Couldn't try locally yet though because my `crc` instance keeps crashing. There might be some failures in the CI since I haven't verified anything locally but this should not block your review.

File: olm-catalog/serverless-operator/manifests/operator_v1alpha1_knativekafka_crd.yaml
Comment (matzew) [reply olm-catalog/serverless-operator/manifests/operator_v1alpha1_knativekafka_crd.yaml:1]: Did my PR for your branch not work @aliok ?

But I can do it later as separate PR 🤷🏻‍♂️

Comment (aliok) [reply olm-catalog/serverless-operator/manifests/operator_v1alpha1_knativekafka_crd.yaml:1]: Yeah, let's do that separately once we merge this PR.

**Review by markusthoemmes — APPROVED** [review]
/lgtm
/approve
/hold

Holding as I'd like to see everything pass and want to doublecheck the logs to contain the new stuff as expected once green.

Great job! :tada:

**Comment by aliok** [comment]
Pushed another commit. Realized the env vars in the makefile were wrong.

**Comment by aliok** [comment]
/test 4.5-upgrade-tests-aws-ocp-45
/test 4.5-upstream-e2e-aws-ocp-45 

```
error after 39m45s. (more info)
Job execution failed: Pod got deleted unexpectedly
```

Thanks @openshift-ci-robot !

**Comment by aliok** [comment]
/test 4.5-operator-e2e-aws-ocp-45

Just retesting one

**Comment by aliok** [comment]
> Yes, one of the build clusters is borked. We're working on moving jobs to the other one.

Response I got from #forum-testplatform

File: test/v1alpha1/knativekafka.go
Comment (markusthoemmes) [inline test/v1alpha1/knativekafka.go:6]: If you have to touch this PR again because of any failures, please remove this dependency. We should be using `fmt.Errorf("blabla: %w, err)` instead of `errors.Wrap`. I have a parallel PR open that fixes other occasions, that's on hold currently.

**Comment by markusthoemmes** [comment]
KnativeKafka doesn't come up

```
{"level":"error","ts":"2020-10-21T14:13:15.586Z","logger":"kubebuilder.controller","msg":"Reconciler error","controller":"knativekafka-controller","request":"knative-eventing/knative-kafka","error":"failed to apply manifest: ServiceAccount \"kafka-ch-controller\" is invalid: [metadata.ownerReferences.apiVersion: Invalid value: \"\": version must not be empty, metadata.ownerReferences.kind: Invalid value: \"\": kind must not be empty]","stacktrace":"github.com/openshift-knative/serverless-operator/knative-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/src/github.com/openshift-knative/serverless-operator/knative-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/openshift-knative/serverless-operator/knative-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/src/github.com/openshift-knative/serverless-operator/knative-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:217\ngithub.com/openshift-knative/serverless-operator/knative-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/go/src/github.com/openshift-knative/serverless-operator/knative-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/openshift-knative/serverless-operator/knative-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/src/github.com/openshift-knative/serverless-operator/knative-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/openshift-knative/serverless-operator/knative-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/src/github.com/openshift-knative/serverless-operator/knative-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/openshift-knative/serverless-operator/knative-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/go/src/github.com/openshift-knative/serverless-operator/knative-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
```

File: test/v1alpha1/knativekafka.go
Comment (markusthoemmes) [reply test/v1alpha1/knativekafka.go:6]: Since this one needs some fixes now, I unheld the other PR, so please make sure to drop this dep.

File: test/v1alpha1/knativekafka.go
Comment (markusthoemmes) [inline test/v1alpha1/knativekafka.go:99]:
```suggestion
		return lastState, fmt.Errorf("knativekafka %s is not in desired state, got: %+v: %w", name, lastState, waitErr)
```

:pray: 

To implement the suggestion above.

**Comment by matzew** [comment]
/test 4.5-operator-e2e-aws-ocp-45

Let;s see if issue persists...

**Comment by aliok** [comment]
Rebased, addressed the `errors.Wrapf()` comment, fixed the order of things in operator e2e tests

**Comment by matzew** [comment]
/lgtm

@markusthoemmes feel free to unhold, after Ali addressed the recent comment - thanks for reviews! :heart:

**Review by markusthoemmes — APPROVED** [review]
/lgtm
/approve

Abfahrt!

**Pull request merged**
